Better error .fileName and .lineNumber default

svaarala opened the pull request:
Use the first traceback entry which actually has a `.fileName` to provide virtual `.fileName` and `.lineNumber`. See #454.

This works better for expressions such as `Duktape.enc(123)` which currently get `undefined` as `.fileName`: the value comes from the `Duktape.enc()` function, which has no `.fileName` property.

Tasks:

- [x] New behavior for traceback-based `.fileName` and `.lineNumber` getters
- [x] New behavior for error augmentation when tracebacks are disabled
- [x] Fix stack trace summarization: previously Duktape would print `[...]` before traceback depth was reached (code didn't account for compilation error or C call site)
- [x] Testcases and test case fixes
- [x] Testcase for empty callstack case
- [x] Testcase for the case where a callstack has entries but none has `.fileName`
- [x] Testcase for the case where a callstack has entries both with and without `.fileName`
- [x] Assert testcase run, with and without tracebacks
- [x] Final code review
- [x] Document the "error file/line attribution" issue in internal error document
- [x] Review and cleanup the internal error document
- [x] Code issues notes on preprocessor defines (somewhat unrelated to this issue)
- [x] Documentation changes (internal docs, website), if necessary
- [x] Releases entry

svaarala commented:
Before the change:

```
duk> try { Duktape.enc(123); } catch (e) { print(e.fileName, e.lineNumber); }
undefined 0
```

After the change:

```
duk> try { Duktape.enc(123); } catch (e) { print(e.fileName, e.lineNumber); }
input 1
```

If one adds a `.fileName` to `Duktape.enc`, that will get blamed:

```
duk> Duktape.enc.fileName = 'dummy.js';
= dummy.js
duk> try { Duktape.enc(123); } catch (e) { print(e.fileName, e.lineNumber); }
dummy.js 0
```

fatcerberus commented:
Off-topic but I didn't realize `print()` accepted multiple parameters, learn something new every day!

svaarala commented:
Yeah it simply joins the values with spaces and appends a newline. The sole exception to this is if print() is called with a single buffer value; in that case the buffer is written as-is to stdout without adding a newline, which allows exact byte output. (Not sure if this is such a good idea, but that's the current behavior.)

svaarala commented:
@fatcerberus If you have a chance, it'd be nice to know if this branch would work for your issue in practice.

fatcerberus commented:
Alright, I'll try this out a bit later.

And thanks for that (presumably undocumented) tip on using a buffer with print() - I was looking for a way a while back to avoid printing the newline but it seemed impossible.

svaarala commented:
Ok, let me know what happens :)

That print() behavior is documented :-) => http://duktape.org/guide.html#builtin-print-alert

fatcerberus commented:
Looks like this works as expected.  I disabled my `errCreate()` handler, then called minisphere's `Abort()` function which has the following signature:

```
Abort(message, callstack_index);
```

`callstack_index` is checked using `duk_require_int()`.  I passed a string instead.  Using Duktape master, the Throw notify indicates "undefined" for the filename, while in this branch I get "script/main.js" line 30, which indeed matches the location of the offending `Abort()`.  So everything seems to be in order. :-)

Now all I need is a version of `duk_error()` that doesn't use `__FILE__` so I can stop calling `duk_error_va_raw()` directly.  I sometimes need to blame errors thrown from C on Ecmascript code... when range-checking function arguments, for example.

svaarala commented:
Do you mean just making the raw variant a part of the public API (#149) or to add some separate variant which doesn't "blame" the call site?

fatcerberus commented:
Implementing #149 would work, too.  Mainly I'm just uncomfortable relying on internal Duktape functions for the long-term, but it's not a terribly pressing concern. :)

svaarala commented:
The current behavior for `duk_error()` is to "blame" the error call site (`__FILE__` / `__LINE__` of the call site). Whether this is a good default or not really depends on the kind of program one is writing. If C code is doing a lot of actual logic, it's natural to pin errors on the C code. But if C code is mostly thin wrappers and such, it's more useful to blame the Ecmascript call site.

Perhaps the `duk_error()` default should be not to blame the C call site?

svaarala commented:
A short overview of how the "blaming" is handled internally:

Functions like `duk_error()` and `duk_push_error_object()` will automatically add the `__FILE__` and `__LINE__` into `_Tracedata`. A flag added to the line number indicates whether this C call should be "blamed" or not when reading back `.fileName` and `.lineNumber`.

When an error is created the "blame or not" flag is carried as part of the error code, e.g.:

```
DUK_ERR_API_ERROR | DUK_ERRCODE_FLAG_NOBLAME_FILELINE
```

This flag is now internal. It'd be trivial to expose it as a part of the API, though I'm not sure that's the best approach for the public API (maybe it is, there are other ORed flags too). I guess API options range from:

- Adding flags to the error code field, which needs no new API calls or arguments in calls
- Adding new API calls for blame / no-blame variants
- Adding (or rather exposing) a raw error throw / create call which has enough flags and parameters to provide flexibility for this and future situations

fatcerberus commented:
Regarding that "noblame" flag, won't that be made redundant by #455?

svaarala commented:
Not as #455 is now: if a C function is "blamed" for the error, `.fileName` and `.lineNumber` will match `__FILE__` and `__LINE__` regardless of whether the Duktape/C function has a `.fileName`or not. All user code `duk_error()` and `duk_push_error_object()` call sites are "blamed" at the moment in this respect.

That blaming behavior could of course be removed entirely - and Duktape/C functions would then only be blamed if the user explicitly added a `.fileName` to the Duktape/C function in question. The `.lineNumber` would still need to be magic'ed from the error call site `__LINE__` somehow.

Maybe that'd be a good change? It would affect the behavior of existing code quite a bit though. While there are no hard promises for `.stack`, `.fileName`, or `.lineNumber`, it's in the grey area whether the change is "surprising" for a minor version.

fatcerberus commented:
Ah, I hadn't realized the tracedata included a separate file and line, I thought it was always taken from the call stack.  But now that I think of it that doesn't make sense, since then `duk_error_raw()` wouldn't work.  Ignore my stupidity. :)

svaarala commented:
I think this issue is worth documenting in the internal errors document, with typical call scenarios and current/preferred behavior. It's kind of hard to figure out the best approach without that.

Luckily we're really only talking about a relatively minor issue here - the traceback always contains the relevant information (including the C call site's `__FILE__` and `__LINE__`, even if it's not blamed). But it'd still be great to have file/line properties be as useful as possible :)

By the way, the `.fileName` of a Duktape/C function (set by the user) may not have anything to do with `__FILE__`: it's entirely possible for the Duktape/C function to call a helper in another C file and the error to be thrown from there. So carrying the `__FILE__`/`__LINE__` combination seems mandatory in any case, but whether to "blame" it is another matter.

Anyway, I'll write up a section on file/line blaming because it's probably worth explaining in a bit more detail.

fatcerberus commented:
Agreed on the above points.  In any case, this pull as it's implemented right now is adequate for my use case, so whatever you decide to do beyond that is fine with me. :)

fatcerberus commented:
@svaarala
Thanks for making all these error message improvements by the way, Duktape's vague error messages were always one of my pet peeves with it compared to other engines.

svaarala commented:
You're welcome, it's a pretty painful but hopefully useful process :)

svaarala commented:
First draft of describing the issue, the Duktape 1.3.0 behavior, and the proposed Duktape 1.4.0 behavior: https://github.com/svaarala/duktape/blob/e86fa305226a42eaff03c69da94b680ee29cae51/doc/error-objects.rst#choosing-filename-and-linenumber-for-an-error

The main issue in the end is that Duktape would need to be able to distinguish between "infrastructure functions" (which are not "blamed" for error file/line) and "user functions" for file/line attribution. Both C and Ecmascript functions may be user and infrastructure functions, so no fixed rule can work for all cases. The same applies to the C call site: some call sites are infrastructure (like all Duktape internal `DUK_ERROR()` sites which are now correctly ignored for file/line purposes) while others are not.

fatcerberus commented:
Found a typo:
`These rules have a first shortcomings.`

I'm assuming that should be `a few shortcomings`.

svaarala commented:
Yeah that was a typo, fixed.

By the way, there are actually *three* potential sources for file/line information; I just remembered the third and added that to the document too:

- Call stack entries
- C call site `__FILE__`/`__LINE__`
- SyntaxError source text file/line

SyntaxErrors source text file/line is included in the stack trace, and is also attributed as the error `.fileName`/`.lineNumber`.

The stack trace also reflects this:

```
duk> try { eval('\n\nfoo='); } catch (e) { print(e.stack); print(e.fileName, e.lineNumber); }
SyntaxError: parse error (line 3)
        input:3                                        <-- file/line of source text (SyntaxError)
        duk_js_compiler.c:3612                         <-- __FILE__ / __LINE__ of DUK_ERROR() call site
        eval  native strict directeval preventsyield   <-- innermost activation, eval() function
        global input:1 preventsyield                   <-- second innermost activation, caller of eval()
input 3
```

svaarala commented:
Internal documents updated, and there's now an implementation for the "no tracebacks" case too. Testcases next.

svaarala commented:
Alright, test case should now cover the new behavior. It should cover the relevant code paths and stack depth corner cases.

fatcerberus commented:
I assume the stack trace for a SyntaxError is simulated?  Because I remember a while back they didn't include file/line info at all (I requested that actually, now that I think of it).

fatcerberus commented:
Regarding the traceback depth, what happens if in the case tracebacks are disabled and the depth limit is exceeded without finding a viable function to blame?  Does Duktape then fall back on the 1.3 behavior?

svaarala commented:
Hmm - not sure what you mean about the SyntaxError?

```
((o) Duktape 1.3.99 (v1.3.0-299-g7c04241)
duk> try { eval('\n\n\nfoo='); } catch (e) { print(e.stack); }
SyntaxError: parse error (line 4)
	input:4
	duk_js_compiler.c:3655
	eval  native strict directeval preventsyield
	global input:1 preventsyield
```

The stack trace includes the source file/line (input:4), the C call site (duk_js_compiler.c:3655), and the actual call stack.

svaarala commented:
When tracebacks are disabled and Duktape can't find a function to blame within the depth limit, the `.fileName` and `.lineNumber` will both be `undefined`. This matches how Duktape would behave with tracebacks enabled. In that case, the runtime lookup for `.fileName` and `.lineNumber` would be handled by the inherited accessors which would consult `_Tracedata` and run out of candidates in the (truncated) tracedata.

fatcerberus commented:
*Is* there a call stack in the case of a syntax error though?  Since it's a compilation error no actual code is being executed (yet).  I assumed that was the cause of SyntaxErrors not including that info in... I think it was v1.1?

svaarala commented:
I just pasted one above :)

svaarala commented:
The call stack in a SyntaxError will reflect the code path which requests compilation. In case of `eval()` it's the callstack leading to the eval.

When you compile functions from C code, it's whatever call stack is present when you call `duk_compile()` or its variants.

svaarala commented:
Syntax errors didn't use to have file/line information (because that's not actually a part of the callstack) but that was added a while ago - but they should've had callstack information pretty much always.

fatcerberus commented:
That's what I meant by simulated - does Duktape generate a dummy callstack entry in the tracedata, or is there a separate code path to handle file/line lookup for syntax errors?  It doesn't matter for user code either way, I just like to reverse engineer stuff. :)

svaarala commented:
Ah ok, you just meant how is that handled internally. There's no substitute for the actual source of course, but it's described in the error document:

- https://github.com/svaarala/duktape/blob/cb16e8bc000c2d81e1846e4aacd3727ca253cf61/doc/error-objects.rst#traceback-format-_tracedata

I added an example to that in this pull.

fatcerberus commented:
I tested this again a bit more thoroughly.   I removed my custom error handler from minisphere code entirely and I still get correct file/line attribution for all `duk_require_xxx()` errors, so this checks out!  I also tested setting a `.fileName` to various C API functions, which had the expected results (this was to ensure minisphere wasn't tampering with the error in some other unforeseen way).

svaarala commented:
Assert runs caught a case with tracebacks disabled and callstack empty, fixed now :) This is pretty ready to go, I'll give it a read-through once more before merging.

fatcerberus commented:
What was the failed assert, if you don't mind me asking?

svaarala commented:
The "no tracebacks" path pushes a file/line to be added to the error object and then writes the properties. The file/line pushing part had separate cases for compilation error, C call site, and scanning the call stack. But if the call stack was empty, nothing got pushed which fouled up the value stack discipline.
